comp/core/autodiscovery/scheduler/controller.go

```diff
@@ -0,0 +1,204 @@
+func (ms *Controller) ApplyChanges(digests []Digest) {
+	for _, configDigest := range digests {
+		ms.queue.Add(workItem{digest: configDigest})
+	}
+}
+
+// Schedule for test only, actual scheduling should be done by the autoconfig.applyChanges
+func (ms *Controller) Schedule(configs []integration.Config) {
+	digests := ms.ConfigStateStore.UpdateDesiredState(integration.ConfigChanges{Schedule: configs})
+	ms.ApplyChanges(digests)
+}
+
+// Unschedule for test only, actual unscheduling should be done by the autoconfig.applyChanges
+func (ms *Controller) Unschedule(configs []integration.Config) {
+	digests := ms.ConfigStateStore.UpdateDesiredState(integration.ConfigChanges{Unschedule: configs})
+	ms.ApplyChanges(digests)
+}
+
+func (ms *Controller) worker() {
+	for ms.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem processes the next work item in the queue
+// Action type will be calculated as following:
+// Current State,   Desired State     Action
+// Unscheduled,     Schedule,         Schedule
+// Unscheduled,     Unschedule,       None
+// Scheduled,       Schedule,         None
+// Scheduled,       Unschedule,       Unschedule
+func (ms *Controller) processNextWorkItem() bool {
+	item, quit := ms.queue.Get()
+	if quit {
+		return false
+	}
+	configDigest := item.(workItem).digest
+	configState, found := ms.ConfigStateStore.GetConfigState(configDigest)
+	if !found {
+		log.Warnf("config %s not found in ConfigStateStore", configDigest)
+		ms.queue.Done(item)
+		return true
+	}
+
+	currentState := Unscheduled
+	desiredState := configState.desiredState
+	configName := configState.config.Name
+	if _, found := ms.scheduledConfigs[string(configDigest)]; found {
+		currentState = Scheduled
+	}
+	if desiredState == currentState {
+		ms.queue.Done(item) // no action needed
+		return true
+	}
+	log.Tracef("Controller starts processing config %s: currentState: %d, desiredState: %d", configName, currentState, desiredState)
+	ms.m.Lock() //lock on activeSchedulers
+	for _, scheduler := range ms.activeSchedulers {
+		if desiredState == Scheduled {
+			//to be scheduled
+			scheduler.Schedule(([]integration.Config{*configState.config})) // TODO: check status of action
+			ms.scheduledConfigs[string(configDigest)] = configState.config
+		} else {
+			//to be unscheduled
+			scheduler.Unschedule(([]integration.Config{*configState.config})) // TODO: check status of action
+			delete(ms.scheduledConfigs, string(configDigest))
+		}
+	}
+	ms.m.Unlock()
+	ms.queue.Done(item)
+	return true
+}
+
+// Stop handles clean stop of registered schedulers
+func (ms *Controller) Stop() {
+	ms.m.Lock()
+	defer ms.m.Unlock()
+	for _, scheduler := range ms.activeSchedulers {
+		scheduler.Stop()
+	}
+	close(ms.stopChannel)
+	ms.queue.ShutDown()
+	ms.started = false
+	ms.scheduledConfigs = make(map[string]*integration.Config)
+}
+
+// Purge removes all scheduled configs and desired states, testing only
+func (ms *Controller) Purge() {
```

Comment:
I don't see a call to `Purge`?

Reply:
Removing it as test doesn't need it now.